timemachine/md/hrex.py

```diff
@@ -0,0 +1,206 @@
+from dataclasses import dataclass
+from typing import Callable, Generic, List, NewType, Optional, Sequence, Tuple, TypeVar
+
+import numpy as np
+from numpy.typing import NDArray
+
+from timemachine.md.moves import Choice, Identity, MetropolisHastingsMove, MonteCarloMove
+from timemachine.utils import batches
+
+_Replica = TypeVar("_Replica")
+
+StateIdx = NewType("StateIdx", int)
+ReplicaIdx = NewType("ReplicaIdx", int)
+
+
+@dataclass
+class NeighborSwapMove(MetropolisHastingsMove[List[_Replica]]):
+    def __init__(self, log_q: Callable[[_Replica, StateIdx], float], s_a: StateIdx, s_b: StateIdx):
+        super().__init__()
+        self.log_q = log_q
+        self.s_a = s_a
+        self.s_b = s_b
+
+    def propose_with_log_q_diff(self, state: List[_Replica]) -> Tuple[List[_Replica], float]:
+        s_a = self.s_a
+        s_b = self.s_b
+        state_ = list(state)
+        state_[s_a], state_[s_b] = state_[s_b], state_[s_a]
+        proposed_state = state_
+
+        r_a = state[s_a]
+        r_b = state[s_b]
+        log_q_diff = self.log_q(r_a, s_b) + self.log_q(r_b, s_a) - self.log_q(r_a, s_a) - self.log_q(r_b, s_b)
+
+        return proposed_state, log_q_diff
+
+
+_Samples = TypeVar("_Samples")
+
+
+@dataclass
+class Hrex(Generic[_Replica]):
+    replicas: List[_Replica]
+    replica_idx_by_state: List[ReplicaIdx]
+
+    @classmethod
+    def from_replicas(cls, replicas: Sequence[_Replica]) -> "Hrex":
+        return Hrex(list(replicas), [ReplicaIdx(i) for i, _ in enumerate(replicas)])
+
+    def sample_replicas(
+        self,
+        sample_replica: Callable[[_Replica, StateIdx], _Samples],
+        replica_from_samples: Callable[[_Samples], _Replica],
+    ) -> Tuple["Hrex[_Replica]", List[_Samples]]:
+
+        samples_by_state = [sample_replica(replica, state_idx) for state_idx, replica in self.state_replica_pairs]
+        replicas_by_state = [replica_from_samples(samples) for samples in samples_by_state]
+
+        replicas = list(self.replicas)
+        for state_idx, replica in enumerate(replicas_by_state):
+            replica_idx = self.replica_idx_by_state[state_idx]
+            replicas[replica_idx] = replica
+
+        return Hrex(replicas, self.replica_idx_by_state), samples_by_state
+
+    def attempt_neighbor_swaps(
+        self,
+        neighbor_pairs: Sequence[Tuple[StateIdx, StateIdx]],
+        log_q: Callable[[ReplicaIdx, StateIdx], float],
+        n_swap_attempts: int,
+    ) -> Tuple["Hrex[_Replica]", List[Tuple[int, int]]]:
+
+        # For each swap attempt, we uniformly choose between the neighboring pairs and an "identity" move that does
+        # nothing. The latter is included as a workaround for slow mixing when neighboring pairs have a near 100% swap
+        # acceptance rate.
+        identity: MonteCarloMove[List[ReplicaIdx]] = Identity()
+        move = Choice([NeighborSwapMove(log_q, s_a, s_b) for s_a, s_b in neighbor_pairs] + [identity])
```

Comment:
rather than mixing in this move, could we augment `neighbor_pairs` with `(i,i) for i in range(n)`?

Reply:
I considered something like this, but wouldn't this mean that ~half of the swap attempts would be no-ops, even if the nearest-neighbor acceptance probabilities are all near 100%?

Reply:
>I considered something like this, but wouldn't this mean that ~half of the swap attempts would be no-ops, even if the nearest-neighbor acceptance probabilities are all near 100%?

Ahh, sorry, good point. That would be a large amount of overhead when `pairs` is "all pairs (i, j) where abs(i - j) == 1"

"no-op" swap attempts would be a smaller fraction of total swap attempts if `pairs` were something like "all pairs (i, j)", or "all pairs (i, j) where abs(i - j) < 10" or something. Possibly something to consider later, but the way you've implemented this for nearest neighbor swaps makes sense.

Reply:
Actually, sorry, I was over-complicating this. Including a single pair `(0,0)` should suffice.

Reply:
> Actually, sorry, I was over-complicating this. Including a single pair (0,0) should suffice.

Nice, I like this suggestion. I was also just thinking that implementing the work-around here breaks the generality of this function, which is supposed to be agnostic to the shape of the graph (e.g. `neighbor_pairs` *could* be all pairs). Will update to remove the mixing with the identity move and just include `(0, 0)` in neighbor pairs in downstream code where we do assume a linear graph.

Reply:
I suppose we could also just remove this, as it's only relevant when K=2, which is probably an uncommon scenario?

Reply:
I opted to just remove this for now, since K=2 seems like an uncommon scenario, and even so I think I'd need more evidence that this is actually an issue. The slow mixing I'd observed that inspired adding this actually turned out to be due to a earlier bug in the code.

Reply:
>I suppose we could also just remove this, as it's only relevant when K=2, which is probably an uncommon scenario?

I'd be in favor of just adding `(0,0)` (if it's not already in the list of neighbor pairs), to ensure that the move is aperiodic even for K=2, and to avoid having to worry about whether the move might be periodic in some unchecked situations for K > 2.

Supporting K = 2 might be useful. (E.g. in cases where we currently jump in one direction from a softened state to a physical state, as in enhanced.py.)

Reply:
>I opted to just remove this for now, since K=2 seems like an uncommon scenario, and even so I think I'd need more evidence that this is actually an issue. The slow mixing I'd observed that inspired this actually turned out to be due to a bug in the code.

Fair enough! I think it should be aperiodic already for K > 2. It might currently be periodic in certain K=2 situations. (E.g. the case you mentioned where accept prob = 100%.)

(Might assert K > 2 or warn if K = 2.)

Reply:
oops, sorry for the crossed wires! I went ahead with an intermediate solution I'm reasonably happy with of manually adding `(0, 0)` to the neighbor pairs in `fe.run_sims_hrex`, which is at a higher level of abstraction and constructs the neighbor pairs given a list of states.

For the lower-level functions, it's currently still up to the caller to ensure that `neighbor_pairs` is constructed in a way to avoid periodic updates when swap acceptance probabilities approach 100% (though could add a warning)

Reply:
In https://github.com/proteneer/timemachine/pull/1128/commits/6a31ac42c0dc57a6b73eef44b19ea22fa062840f, modified to add an identity move if and only if K=2, since I don't think there is a potential issue with aperiodicity for K>2.

Reply:
>I don't think there is a potential issue with aperiodicity for K>2.

👍 , I think it should be aperiodic already for K > 2.